### PR TITLE
feat: json functions 

### PR DESCRIPTION
**Pull request by mnfienen**
JSON functions to instatiate PrmsParameters object from JSON and to save or load parameter dictionary as JSON.

Needs tests

### CODE REVIEW

**Comment by jmccreight** [comment]
I merged a lingering PR into to upstream:main and then merged upstream:main into your fork:branch. the RTD build succeeds now.

I wont merge without a test. I think the test is simple read a myparams.param file (a single domain should suffice) to a parameter object, write it to json, read the json back in and check that the read json gives an identical parameter object. You'll have to check the parts or define equality for the class.

**Comment by mnfienen** [comment]
kewl @jmccreight - that test idea was what I was planning. I just wasn't certain there was a reliable location for a `params` file to work with. I see this `domain` object in the `test_parameter_read.py` stuff but couldn't sort out what that is. If you can point me to a `params` file in the testing suite I'll dial in the test.

**Comment by jmccreight** [comment]
now that I've thought about it, let me propose something. I'll push it to your fork:branch

File: pynhm/utils/parameters.py [hunk 1]
```diff
@@ -218,6 +244,34 @@ def nhm_coordinates(self) -> dict:
...
 
+    def parameters_to_json(self, json_filename):
+        """write the parameters dictionary out to a json file"""
+        json.dump(
+            self.parameters,
+            open(json_filename, "w"),
+            indent=4,
+            cls=JSONParameterEncoder,
+        )
+
```
Comment (jmccreight) [inline pynhm/utils/parameters.py:255]: I kinda like `return` at the end of functions, if extra pedantic `return None`. I like them visually

**Comment by jmccreight** [comment]
I'm about to push my rough draft, cause I got to put it down for now. It's close but there are checks that are not passing and it's not obvious to me why. Here's where I get to

Screen Shot 2023-01-20 at 9 38 43 AM

**Comment by mnfienen** [comment]
oy - that's going to be a little tricky since I relied on the constructor to make those dimensions. But, I'll take a look

**Comment by mnfienen** [comment]
oh - actually I see the problem. I will push up a fix later today

**Comment by jmccreight** [comment]
good, cause it's not obvious to me! LMK if there's anything i can help on. 

The only additional thing I'd say is that it'd be good to test `update_parameters_from_json` as well. Am I clear on the intent:? given an existing parameter object, `update_parameters_from_json` would slurp up a json and update the values in the parameter object with those found in the json, which may only be a subset. Will you actually use that since you have to interact through files anyway?

**Comment by jmccreight** [comment]
BTW, we can go over conceptually what's going on here sometime. the domain is just a parameterized pytest fixture at the conftest level, it gives all the necessary inputs to run a given model domain for multiple model domains. Now it's ALL clear!

**Comment by jmccreight** [comment]
I had to fix the default discovery of param dimension names which wasnt getting exercised. I enhanced Joe's early wild guess code for param dimension names with the metadata module that since became available. This was a long time coming and it works with only the hitch that 4 previously unidentified parameters needed added to the static/metadata/parameters.yaml, i think 3 were for SNTMP and the other was a precip threshold, none appeared in the existing code base.

**Pull request merged**